"""
TXF Streaming Consumer (Kafka -> Console Inspector)
---------------------------------------------------
Description: 
    å¾ Kafka æ¥æ”¶ TXF Tick èˆ‡ BidAsk æ•¸æ“š (Protobuf æ ¼å¼)ã€‚
    æ¡ç”¨ Batch Consume æ¨¡å¼ï¼Œè¨ˆç®—ä¸¦æ‹†è§£å»¶é² (Latency) ç‚ºï¼š
    1. å¤–éƒ¨å»¶é² (ç¶²è·¯/Queue Lag)
    2. å…§éƒ¨å»¶é² (Python è§£æèˆ‡è™•ç†é–‹éŠ·)
    
Architecture:
    - Deserialization: Google Protobuf
    - Transport: confluent-kafka (Batch Consume)
    - Precision: Nanosecond (ns) based time calculation

Author: Garrett & Gemini
Last Updated: 2025-11-28
"""

import sys
import time
from datetime import datetime

# --- Third-party Imports ---
from confluent_kafka import Consumer, KafkaError

# --- Local Imports ---
from . import txf_data_pb2
from .config import (
    KAFKA_BOOTSTRAP_SERVERS, 
    TICK_TOPIC, 
    BIDASK_TOPIC
)

# ==========================================
# Global Constants
# ==========================================

SCALE = 10000.0
GROUP_ID = 'txf-console-inspector'

# ==========================================
# Helper Functions
# ==========================================

def to_decimal(scaled_int: int) -> float:
    """å°‡ int64 (x10000) é‚„åŸå› floatã€‚"""
    return scaled_int / SCALE

def format_list(data_list, is_price=False) -> str:
    """å°‡åˆ—è¡¨æ ¼å¼åŒ–ç‚ºæ˜“è®€å­—ä¸² (é¡¯ç¤ºå‰3æª”)ã€‚"""
    if not data_list: return "[]"
    if is_price:
        items = [str(to_decimal(x)) for x in data_list[:3]]
        return f"[{', '.join(items)} ...]"
    items = [str(x) for x in data_list[:3]]
    return f"[{', '.join(items)} ...]"

# ------------------------------------------------------------
# ğŸ“¦ æ‰“å°å‡½æ•¸ (æ ¼å¼åŒ–è¼¸å‡º)
# ------------------------------------------------------------
def print_event_summary(topic, data, transport_latency_ms, internal_latency_ms, msg_count):
    """
    æ ¹æ“š topic é¡å‹æ‰“å°æ ¼å¼åŒ–çš„ Logã€‚
    - transport_latency: ç¶²è·¯/Queue å»¶é²
    - internal_latency: Python è™•ç†å»¶é²
    """
    
    total_latency = transport_latency_ms + internal_latency_ms
    
    if topic == TICK_TOPIC:
        event_dt = datetime.fromtimestamp(data.timestamp_ms / 1000.0).strftime('%H:%M:%S.%f')[:-3]
        type_str = {1: "å¤–ç›¤", 2: "å…§ç›¤"}.get(data.tick_type, "æœªçŸ¥")

        print(f"\nâš¡ [TICK] {data.code} @ {event_dt} (ç¸½å»¶é² {total_latency:.1f}ms | æ‰¹æ¬¡: {msg_count})")
        print(f"   â”œâ”€ å¤–éƒ¨å‚³è¼¸: {transport_latency_ms:.1f}ms (Queue Lag)")
        print(f"   â”œâ”€ å…§éƒ¨è™•ç†: {internal_latency_ms:.3f}ms (è§£æé–‹éŠ·)") 
        print(f"   â”œâ”€ æˆäº¤åƒ¹: {to_decimal(data.close):.0f} | å–®é‡: {data.volume}")
        print(f"   â””â”€ å…§å¤–ç›¤: {data.tick_type} ({type_str})")

    elif topic == BIDASK_TOPIC:
        event_dt = datetime.fromtimestamp(data.timestamp_ms / 1000.0).strftime('%H:%M:%S.%f')[:-3]
        bid1 = to_decimal(data.bid_price[0]) if data.bid_price else 0
        ask1 = to_decimal(data.ask_price[0]) if data.ask_price else 0
        
        print(f"\nğŸ“Š [BID/ASK] {data.code} @ {event_dt} (ç¸½å»¶é² {total_latency:.1f}ms | æ‰¹æ¬¡: {msg_count})")
        print(f"   â”œâ”€ å¤–éƒ¨å‚³è¼¸: {transport_latency_ms:.1f}ms (Queue Lag)")
        print(f"   â”œâ”€ å…§éƒ¨è™•ç†: {internal_latency_ms:.3f}ms (è§£æé–‹éŠ·)") 
        print(f"   â””â”€ è²·ä¸€: {bid1:.0f} | è³£ä¸€: {ask1:.0f}")

# ==========================================
# Main Execution
# ==========================================

def main():    
    # --- 1. åˆå§‹åŒ– Consumer ---
    print("ğŸ”§ åˆå§‹åŒ– Kafka Consumer (æ™‚é–“å»¶é²åˆ†ææ¨¡å¼)...")
    
    conf = {
        'bootstrap.servers': KAFKA_BOOTSTRAP_SERVERS,
        'group.id': GROUP_ID,
        'auto.offset.reset': 'latest',
        'enable.auto.commit': True
    }

    try:
        consumer = Consumer(conf)
        consumer.subscribe([TICK_TOPIC, BIDASK_TOPIC])
        
        print(f"ğŸš€ ç›£æ§å•Ÿå‹• | Group: {GROUP_ID}")
        print(f"ğŸ“¡ ç›£è½ Topics: {TICK_TOPIC}, {BIDASK_TOPIC}")
        print( "â³ ç­‰å¾…è³‡æ–™ä¸­... (æŒ‰ Ctrl+C åœæ­¢)")
        print( "-" * 60)
        
    except Exception as e:
        print(f"âŒ Consumer åˆå§‹åŒ–å¤±æ•—: {e}")
        sys.exit(1)

    # --- 2. é€²å…¥ç›£æ§è¿´åœˆ ---
    try:
        # [è¨ˆæ™‚è®Šæ•¸]
        deadline = time.time() + 1.0 # æ§åˆ¶æ¯ç§’è¼¸å‡ºä¸€æ¬¡ (é™æµ)

        while True:
            # (A) é‡ç½®ç‹€æ…‹ (æ¯è¼ªé–‹å§‹æ™‚ï¼Œå‡è¨­æ²’æœ‰ Tick/BA)
            latest_tick_msg = None
            latest_ba_msg = None

            # (B) æ‰¹æ¬¡æŠ“å– (0.1s timeoutï¼Œç¢ºä¿ Consumer ä¸æœƒç©ºè½‰)
            msgs = consumer.consume(num_messages=500, timeout=0.1)
            
            # ã€é—œéµé» 1: T_Arrivalã€‘ Python æ‹¿åˆ°é€™æ‰¹è³‡æ–™çš„æ™‚é–“é»
            # é€™æ˜¯å»¶é²è¨ˆç®—çš„ç¬¬äºŒå€‹åŸºæº–é» (è³‡æ–™æŠµé”æ‡‰ç”¨ç¨‹å¼çš„æ™‚é–“)
            arrival_time_ns = time.time_ns()
            
            if not msgs: 
                continue
            
            # --- 3. éæ­·æ‰¹æ¬¡ (In-Batch Tracking) ---
            valid_msgs = [m for m in msgs if not m.error()]
            valid_msgs_count = len(valid_msgs)
            if valid_msgs_count == 0: continue
            
            for msg in valid_msgs:
                topic = msg.topic()
                # è¿½è¹¤æ¯å€‹ Topic æœ€å¾Œæ”¶åˆ°çš„é‚£æ¢è¨Šæ¯ (æœ€æ–°)
                if topic == TICK_TOPIC:
                    latest_tick_msg = msg
                elif topic == BIDASK_TOPIC:
                    latest_ba_msg = msg

            # --- 4. å®šæ™‚æ‰“å°é‚è¼¯ (æ¯ç§’ä¸€æ¬¡) ---
            if time.time() >= deadline:
                
                # (!! é—œéµï¼šè§£æèˆ‡è¨ˆç®—å¿…é ˆåœ¨ if è£¡é¢ï¼Œé¿å…ç„¡æ„ç¾©çš„ CPU é‹ç®— !!)
                
                # è™•ç† TICK
                if latest_tick_msg:
                    tick_data = txf_data_pb2.Tick()
                    
                    # [T_Ready Start] è¨ˆç®—è§£æè€—æ™‚çš„èµ·é»
                    start_parse_ns = time.time_ns()
                    tick_data.ParseFromString(latest_tick_msg.value())
                    
                    # [T_Ready End] è¨ˆç®—è§£æè€—æ™‚çš„çµ‚é»
                    parse_done_ns = time.time_ns()
                    
                    # è¨ˆç®—å»¶é² (å–®ä½æ›ç®—ç‚º ms)
                    # å¤–éƒ¨å‚³è¼¸ = (T_Arrival) - (T_Event)
                    transport_latency = (arrival_time_ns / 1_000_000) - tick_data.timestamp_ms
                    
                    # å…§éƒ¨è™•ç† = (T_Ready) - (T_Arrival)
                    internal_latency = (parse_done_ns - start_parse_ns) / 1_000_000
                    
                    print_event_summary(TICK_TOPIC, tick_data, transport_latency, internal_latency, valid_msgs_count)

                # è™•ç† BIDASK
                if latest_ba_msg:
                    ba_data = txf_data_pb2.BidAsk()
                    
                    # [T_Ready Start]
                    start_parse_ns = time.time_ns()
                    ba_data.ParseFromString(latest_ba_msg.value())
                    
                    # [T_Ready End]
                    parse_done_ns = time.time_ns()
                    
                    transport_latency = (arrival_time_ns / 1_000_000) - ba_data.timestamp_ms
                    internal_latency = (parse_done_ns - start_parse_ns) / 1_000_000
                    
                    print_event_summary(BIDASK_TOPIC, ba_data, transport_latency, internal_latency, valid_msgs_count)
                
                print("-" * 40)
                deadline = time.time() + 1.0 # (é‡è¨­ä¸‹ä¸€æ¬¡çš„æ‰“å°æ™‚é–“)

    except KeyboardInterrupt:
        print("\nğŸ‘‹ æ”¶åˆ°åœæ­¢è¨Šè™Ÿï¼Œç›£æ§çµæŸã€‚")
        
    finally:
        print("â³ æ­£åœ¨é—œé–‰ Consumer...")
        consumer.close()
        print("âœ… Consumer å·²é—œé–‰ã€‚")

if __name__ == '__main__':
    main()